File: artifacts/api-server/src/lib/sentry-init.ts
import * as Sentry from "@sentry/node";
import { getActiveConfig } from "./integrations";
import { logger } from "./logger";

let initialized = false;

/**
 * Initialise Sentry from the integrations panel config (if enabled).
 *
 * Why read from DB instead of env var: the whole point of the integrations
 * framework is that the admin pastes the DSN once in /integrations, toggles
 * it on, and it just works — no redeploy, no env-var management.
 *
 * Sentry's SDK MUST be initialised before any module that might throw, so
 * this is called at the very top of index.ts (before app import).
 *
 * Safe to call multiple times — second call is a no-op.
 */
export async function initSentryFromIntegration(): Promise<boolean> {
  if (initialized) return true;
  try {
    const cfg = await getActiveConfig("sentry");
    if (!cfg || !cfg["dsn"]) {
      logger.info("Sentry not configured/enabled — skipping init");
      return false;
    }
    const tracesSampleRate = Number(cfg["traces_sample_rate"] ?? "0");
    Sentry.init({
      dsn: cfg["dsn"],
      environment: cfg["environment"] || "production",
      tracesSampleRate: Number.isFinite(tracesSampleRate) ? tracesSampleRate : 0,
      // Strip query strings and auth tokens from URLs sent to Sentry.
      beforeSend(event) {
        if (event.request?.headers) {
          delete event.request.headers["authorization"];
          delete event.request.headers["cookie"];
          delete event.request.headers["x-bot-api-key"];
        }
        return event;
      },
    });
    initialized = true;
    logger.info({ env: cfg["environment"] }, "Sentry initialised from integration");
    return true;
  } catch (err) {
    logger.error({ err }, "Sentry init failed");
    return false;
  }
}

export { Sentry };

File: artifacts/books-bot/src/sentry_init.py
"""
Sentry bootstrap shared by all Python bots.

Reads DSN from the API's /internal/runtime/sentry endpoint so the admin
only has to paste the DSN once in /integrations — no duplicate env vars
across 3 bot processes. Silent no-op if Sentry isn't enabled, so dev
environments without Sentry still work.

Usage: call init_sentry(bot_name="mother-bot") at the very top of bot.py,
BEFORE importing any module that might raise.
"""
import os
import logging

import httpx
import sentry_sdk

logger = logging.getLogger(__name__)


def init_sentry(bot_name: str) -> bool:
    """
    Init Sentry from the API integrations panel.
    Returns True if successfully initialised, False otherwise.
    Never raises — Sentry must never break the bot.
    """
    api_url = os.getenv("MOTHER_API_URL", "http://localhost:80/api").rstrip("/")
    api_key = os.getenv(f"{bot_name.upper().replace('-', '_')}_API_KEY") \
              or os.getenv("MOTHER_BOT_API_KEY", "")
    if not api_key:
        logger.info("Sentry init skipped: no bot API key in env")
        return False
    try:
        resp = httpx.get(
            f"{api_url}/internal/runtime/sentry",
            headers={"X-Bot-Api-Key": api_key},
            timeout=5.0,
        )
        if resp.status_code != 200:
            logger.info("Sentry init skipped: API returned %d", resp.status_code)
            return False
        data = resp.json()
        if not data.get("enabled"):
            logger.info("Sentry not enabled in admin panel — skipping")
            return False
        dsn = data.get("dsn")
        if not dsn:
            return False
        sentry_sdk.init(
            dsn=dsn,
            environment=data.get("environment", "production"),
            traces_sample_rate=float(data.get("traces_sample_rate", 0) or 0),
            # Tag every event with which bot raised it.
            release=bot_name,
            before_send=_scrub,
        )
        sentry_sdk.set_tag("bot", bot_name)
        logger.info("Sentry initialised for %s", bot_name)
        return True
    except Exception as e:  # noqa: BLE001
        logger.warning("Sentry init failed (non-fatal): %s", e)
        return False


def _scrub(event, hint):  # type: ignore[no-untyped-def]
    """Strip auth headers before sending to Sentry."""
    req = event.get("request") or {}
    headers = req.get("headers") or {}
    for h in ("authorization", "x-bot-api-key", "cookie"):
        headers.pop(h, None)
        headers.pop(h.title(), None)
    return event
